[Week8] API 문서 만들기, Docker image 만들기

JIY0UNG opened the pull request:
Spring REST Docs를 사용하여 ProductController에 대한 API 문서를 작성해보았습니다!

JIY0UNG committed:
Asciidoctor를 사용하기 위한 설정을 하라

- Asciidoctor를 사용하여 /generated-snippets 하위에 생성된 파일들을 HTML로 변환하기 위해 spring-restdocs-asciidoctor에 대한 의존성을 추가하였습니다.
- 테스트 코드를 문서화해주는 REST Docs를 사용하기 위해spring-restdocs-mockmvc에 대한 의존성을 추가하였습니다.

JIY0UNG committed:
Product 컨트롤러 테스트 코드에 REST Docs 문서명을 지정하는 코드를 추가하라

JIY0UNG committed:
상품에 대한 REST Docs를 작성하라

JIY0UNG commented:
처음에 강의에 나온대로
```java
plugins {
    ...
    // Asciidoctor
    id 'org.asciidoctor.convert' version '2.4.0'
}
...
dependencies {
    ...
    // Spring REST Docs
    asciidoctor 'org.springframework.restdocs:spring-restdocs-asciidoctor'
    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
}
...
asciidoctor {
    dependsOn test
}
```
이렇게 설정한 후 터미널에서 `gradlew asciidoc` 명령어를 실행하니 아래와 같은 에러가 발생했습니다.

```bash
> Task :app:asciidoctor FAILED
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.codehaus.groovy.reflection.CachedClass (file:/C:/Users/HJY/.gradle/wrapper/dists/gradle-6.7-bin/efvqh8uyq79v2n7rcncuhu9sv/gradle-6.7/lib/groovy-all-1.3-2.5.12.jar)
 to method java.lang.Object.finalize()
WARNING: Please consider reporting this to the maintainers of org.codehaus.groovy.reflection.CachedClass
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Exception in thread "main" org.jruby.exceptions.MainExitException: unknown encoding name - x-windows-949
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':app:asciidoctor'.
> Process 'command 'C:\Program Files\Java\jdk-15.0.2\bin\java.exe'' finished with non-zero exit value 1
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 6s
5 actionable tasks: 3 executed, 2 up-to-date
```

--stacktrace로 자세한 에러 메시지를 읽어보기도 했으나, 무엇이 문제인지 정확하게 파악이 어려웠습니다.

에러 메시지 중간에 gradle이 나오는 걸 보고 버전 문제인가 싶어서 이에 대해 찾아보았습니다.
Gradle 버전을 낮추거나 Asciidoctor 플러그인의 버전을 높이면 해결이 된다는 글을 발견해서 아래와 같이 asciidoctor의 plugin 설정을 변경하니 에러 없이 빌드가 잘 되었습니다.

```java
plugins {
    ...
    // Asciidoctor
    id 'org.asciidoctor.jvm.convert' version "3.3.2"
}

configurations {
    ...
    asciidoctorExtensions
}
...
dependencies {
    ...
    // Spring REST Docs
    asciidoctorExtensions 'org.springframework.restdocs:spring-restdocs-asciidoctor'
    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
}
...
asciidoctor {
    dependsOn test
    configurations 'asciidoctorExtensions'
}
```

그런데 에러 메시지 중 이 부분을 보면
> What went wrong:
> Execution failed for task ':app:asciidoctor'.
> Process 'command 'C:\Program Files\Java\jdk-15.0.2\bin\java.exe'' finished with non-zero exit value 1

자바 버전이 안 맞아서 발생한 에러인가 싶기도 합니다.

그래서 지금 제가 저 에러 메시지에 맞는 적절한 대처를 해서 해결이 된 건지, 어쩌다 얻어걸린 건지 잘 모르겠는데, 한번 봐주실 수 있나요??

가끔 이것저것 고치다가 갑자기 해결되는 경우가 종종 있는데,
'이런 에러가 발생하면 → 이런 행동을 해야 한다'를 확실히 짚고 넘어가고 싶어서 이렇게 질문 드려봅니다

ssisksl77 commented:
어쩌면 이리저리 검색해서 확인하신 내용은 올려주신 예외에 힌트로 이미 존재하는 것 같군요!

`Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.`

ssisksl77 commented on app/build.gradle:
```diff
@@ -78,6 +82,10 @@ dependencies {
+
+    // Spring REST Docs
+    asciidoctorExtensions 'org.springframework.restdocs:spring-restdocs-asciidoctor'
+    testImplementation 'org.springframework.restdocs:spring-restdocs-mockmvc'
```
이것을 추가하면 무엇을 사용할 수 있는 건가요?! 커밋메시지(df0f5e8d98f12a90585e035abb075b0f1e626ab9)에 추가해보세요!

JIY0UNG replied:
- spring-restdocs-asciidoctor에 대한 의존성을 추가하면, /generated-snippets 하위에 생성된 파일들을 HTML로 변환해주는 Asciidoctor를 사용할 수 있게 됩니다.
- spring-restdocs-mockmvc에 대한 의존성을 추가하면, 테스트 코드를 문서화해주는 REST Docs를 사용할 수 있게 됩니다.

커밋메시지에 위 내용을 추가하였습니다!

ssisksl77 replied:
> * spring-restdocs-mockmvc에 대한 의존성을 추가하면, 테스트 코드를 문서화해주는 REST Docs를 사용할 수 있게 됩니다.

REST Docs 를 사용하는 건 무엇을 말하는 걸까요? 좀 더 정확히 무엇을 사용하기 위함일까요?!
관련 정보를 함께 공유해주세요!

JIY0UNG replied:
Spring REST Docs는 Asciidoctor로 작성된 문서와 Spring MVC Test로 자동 생성된 스니펫을 결합하여 RESTful 서비스를 문서화할 수 있도록 해줍니다.
Spring REST Docs를 사용하면 HTTP 요청과 응답으로 문서를 만들 수 있는데, 서비스 구현으로부터 문서가 분리되어 있어서 문서를 다시 수정하지 않아도 구현을 변경할 수 있습니다.

ssisksl77 replied:
좀 더 정확히 어떤 객체를 사용하기 위함인가요?!

ssisksl77 replied:
참고로 사용하신 `@AutoConfigureRestDocs` 는 `spring-restdocs-mockmvc` 에 있지 않습니다! 

`org.springframework.restdocs:spring-restdocs-mockmvc` 의존성은 REST Docs에서 무엇을 담당하는 것일까요?!

의존성을 하나하나 추가할 때마다 이것을 추가하면 나는 무엇을 할 수 있게 되는거지? 정확히 무엇을 써서 무엇을 할 수 있게 되는거지? 를 메시지에 적어주셨으면 하는 마음에 질문드립니다!^^

JIY0UNG replied:
테스트 코드를 스니펫으로 만들 때 `document("get-products")` 와 같이 document 메서드를 사용하는데, 
`spring-restdocs-mockmvc`에 대한 의존성을 추가해줘야 MockMvcRestDocumentation 클래스의 document 메서드를 사용할 수 있습니다.

`@AutoConfigureRestDocs`는 `spring-boot-test` 의존성에 의해 사용 가능한 것이었군요!

ssisksl77 replied:
맞아요! 추가로 MockMvc 라는 객체를 사용하기 위해서 사용하는 것이죠! 

여기서 한가지 더 의문이 생깁니다! RESTDocs 을 만들 때 테스트를 작성하는 것은 필수인데 왜 의존성을 따로 넣어야 하는 것일까요?

구현체가 하나가 아닌 것이 아닐까요?!
다른 구현체가 있다면 그것에 대해 알아보도록해요!

미래에 어떤 구현체를 선택해서 PR을 올리면 '왜 다른 구현체를 사용하지 않고 이 구현체를 선택했는지' 물어보는 분들이 많을거에요!

JIY0UNG commented:
> 어쩌면 이리저리 검색해서 확인하신 내용은 올려주신 예외에 힌트로 이미 존재하는 것 같군요!
> 
> `Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.`

저 메시지는 강의에서 아샬님이 빌드했을 때도 똑같이 나오는데, 아샬님은 아래 사진처럼 빌드에 성공하시더라구요.
![제목 없음](https://user-images.githubusercontent.com/98736689/175045441-5eaaf271-6515-460c-9588-7b02a1df345e.png)
그래서 버전 문제는 아닌건가, 하는 의문이 들었습니다.

---

방금 에러 메시지를 다시 잘 살펴봤는데 이번에는 아래 메시지가 눈에 들어와서,
> Exception in thread "main" org.jruby.exceptions.MainExitException: unknown encoding name - x-windows-949

다시 에러가 발생하는 상태로 되돌린 후 인텔리제이의 모든 인코딩 설정을 UTF-8로 변경하고 빌드 해 봤지만 똑같은 에러가 발생하네요..🤔

제가 잘 못 찾는 건지, 에러 메시지를 여러 번 읽어봐도 뭔가 '아 이 부분을 고쳐야겠구나! '라는 게 명확하게 보이지 않아서 혼란스러운 것 같습니다..

ssisksl77 commented:
> ```shell
> org.jruby.exceptions.MainExitException: unknown encoding name - x-windows-949
> FAILURE: Build failed with an exception.
> ```

던져진 예외에 있는 인코딩 이름 windows-949 는 윈도우에서 사용하는 거죠!

인텔리제이의 모든 인코딩을 UTF-8로 바꿨더라도 동일한 예외를 던진다면 실제로 디버깅을 해서 어디에서 encoding을 설정하는지 확인해볼거 같아요!

ssisksl77 commented:
실제 강의와 다른 이유는 강의에선 맥북을 사용하면서 windows-949인코딩을 사용하고 있지는 않을거에요!

JIY0UNG commented:
> 실제 강의와 다른 이유는 강의에선 맥북을 사용하면서 windows-949인코딩을 사용하고 있지는 않을거에요!

강의에서는 맥북을 쓰고 있다는 사실을 잊고 있었네요!

Mac OS는 유닉스 환경이라 기본적인 인코딩 설정이 UTF-8로 되어있지만,
Windows는 기본 인코딩 설정이 MS-949이기 때문에 build.gradle 파일을 빌드할 때 이런 인코딩 오류가 난다는 것을 알게 되었습니다.

그래서 Gradle 빌드에 대한 별도의 인코딩 설정을 어떻게 해야하는지 찾아보았고,
build.gradle 파일에 아래와 같이 인코딩 설정 코드를 추가함으로써
asciidoctor 플러그인의 버전을 변경하지 않고도 빌드에 성공할 수 있었습니다.
```java
compileJava.options.encoding = 'UTF-8'
tasks.withType(JavaCompile) {
    options.encoding = 'UTF-8'
}
```

---

처음에는 에러 메시지를 보고도 대체 어디가 문제라는 건지 파악이 안됐었는데,
> Exception in thread "main" org.jruby.exceptions.MainExitException: unknown encoding name - x-windows-949
FAILURE: Build failed with an exception.

예외가 던져졌다는 것은 즉 여기서 에러가 났다는 거니까
이 부분이 에러 메시지의 핵심이라고 보면 되겠군요!

JIY0UNG commented:
> 실제로 디버깅을 해서 어디에서 encoding을 설정하는지 확인해볼거 같아요!

그런데 혹시 이런 파일은( build.gradle) 디버깅을 어떻게 하는건가요??
일반 자바 코드처럼 중단점 찍고 디버깅을 하나요??

JIY0UNG committed:
Asciidoctor 플러그인 버전을 되돌리고, 컴파일 인코딩 설정을 추가하라

- 기존에 Gradle build 시 발생했던 에러의 원인을 알 수 없어서 Asciidoctor 플러그인의 버전을 높이는 것으로 임시적인 해결을 했었습니다.
후에 에러의 원인이 인코딩 설정에 있다는 것을 알게 되었고, 컴파일 시 UTF-8로 인코딩 하도록 설정하는 코드를 추가해주었습니다.
- 위의 에러가 해결되어 더 높은 버전의 Asciidoctor 플러그인을 사용할 이유가 딱히 없기 때문에, 원래 버전으로 되돌렸습니다.

JIY0UNG committed:
Session 컨트롤러 테스트 코드에 REST Docs 문서명을 지정하는 코드를 추가하라

JIY0UNG committed:
생성된 스니펫으로 로그인에 대한 REST Docs를 작성하라

JIY0UNG committed:
User 컨트롤러 테스트 코드에 REST Docs 문서명을 지정하는 코드를 추가하라

JIY0UNG committed:
생성된 스니펫으로 회원에 대한 REST Docs를 작성하라

ssisksl77 commented:
> > 실제로 디버깅을 해서 어디에서 encoding을 설정하는지 확인해볼거 같아요!
> 
> 그런데 혹시 이런 파일은( build.gradle) 디버깅을 어떻게 하는건가요?? 일반 자바 코드처럼 중단점 찍고 디버깅을 하나요??

전 현재까지 build.gradle 로 큰 문제를 일으켰던 경험이 없어서 `--debug` 정도로 출력된 정보로 
 - 어떤 매개변수와 함께 실행되었는지
 - 예외가 일어난 객체는 어디인지 
정도 확인하고 해당 소스코드를 찾아보는 것만으로도 해결이 되었던 것 같아요!

하지만 만약에 제대로 테스트하고 디버그 하고 싶다면 아래 링크가 힌트가 되길 바랄게요!
https://www.jetbrains.com/help/idea/run-debug-gradle.html

ssisksl77 commented:
https://docs.spring.io/spring-restdocs/docs/current/reference/html5/#documenting-your-api-request-response-payloads

위 링크를 참조해서 요청응답 페이로드를 만들어보아요!

JIY0UNG committed:
Javadoc을 작성하라

JIY0UNG commented:
코드 설명을 위한 Javadoc을 추가했습니다!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -24,19 +27,45 @@ public ProductService(
     }
 
+    /**
+     * 상품을 생성하고, 생성된 상품을 반환한다.
```
앞부분은 제거해도 될 것 같아요!^^

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -45,6 +74,13 @@ public Product updateProduct(Long id, ProductData productData) {
     }
 
+    /**
+     * id로 조회된 상품을 삭제하고, 삭제된 상품을 반환한다.
```
'id로 조회된 상품을 삭제하고,' 에서 id로 조회한다는 것은 구현을 너무 노출한 건 아닐까요?! 
한 번 고민해보아요!

ssisksl77 commented on app/src/main/java/com/codesoom/assignment/application/ProductService.java:
```diff
@@ -53,6 +89,13 @@ public Product deleteProduct(Long id) {
         return product;
     }
 
+    /**
+     * id로 상품을 찾고, 찾은 상품을 반환한다.
+     * 
+     * @param id 상품의 id
+     * @return 조회된 상품
+     * @throws ProductNotFoundException id와 일치하는 상품이 조회되지 않은 경우에 발생
+     */
```
다른 문서도 비슷하지만 구체적인 설명이 있는 것은 좋지만 구체적인 구현이 설명되어있는 것은 오히려 가독성이 떨어질 수 있어요!

만약에 파라미터가 1개 더 추가되었다고 합니다! email이 생겼다고 해보죠!
그러면, `id,status로 상품을 찾고 ...` 가 되고
`id,status과 일치하는 상품이 조회되지 않은 경우에 발생` 
이렇게 바뀐다고 합시다!

만약에 여기서 N개의 인자가 추가되었다고 해봐요! 그렇다면 문서에 N개를 다 나열해본다고 생각해봅시다! 
나열된 N개가 문서를 읽는데 방해가 될 수도 있겠죠. 그리고 찾다와 조회하다처럼 같은 뜻을 두 개의 단어로 쓰면 혹여 다른 뜻이 있는걸까? 고민할 수도 있어요!

예를들면 
```
요청한 인자로 조회된 상품을 반환한다.

@param id 상품 id
@return 조회된 상품
@throws ProductNotFoundException 요청한 상품이 조회되지 않은 경우
```

JIY0UNG replied:
> 다른 문서도 비슷하지만 �구체적인 설명이 있는 것은 좋지만 구체적인 구현이 설명되어있는 것은 오히려 가독성이 떨어질 수 있어요!

저번 주 과제에서도 같은 리뷰를 받아서, 이번에는 구현에 대한 설명을 최대한 뺐다고 생각했는데 아니었군요😂
'구체적인 구현을 설명하지 않는 것'을 좀 더 쉽게 생각해보면, '반환값을 제외한 나머지 요소들(파라미터나 예외)을 구체적으로 명시하지 않는 것'이라고 보면 될까요??

JIY0UNG replied:
> 그리고 찾다와 조회하다처럼 같은 뜻을 두 개의 단어로 쓰면 혹여 다른 뜻이 있는걸까? 고민할 수도 있어요!

정말 그럴 수 있겠네요, 이런 식으로 두 개의 단어가 사용된 곳은 모두 하나의 단어로 수정하겠습니다!

ssisksl77 replied:
인자와 예외가 javadoc에 없어져야 하는 것은 아니에요!

인자와 예외는 javadoc 아래부분에 이미 다 설명이 되어있지요!

그렇기에 내용에 적어야 할 필요가 없는 해당 함수에 대한 역할을 이해하기 위한 정보만 있으면 된다고 생각해요

ssisksl77 replied:
좀 더 이야기 하자면 현재 만드신 함수는 간단하기 때문에 어떻게 문서를 작성해도 상관이 없어요! 절대 이해를 못할리 없겠죠!

하지만 만약 구현이 복잡해질 경우의수가 많을 것이라고 예상되는 함수라고 상상을 해보고 이야기를 나누는 것입니다!

만약에 그렇다면 함수가 더 복잡해면 복잡해질 수록 문서에 A인자를 어떻게 하고 B인자를 어떻게 하고 같은 모든 것을 적게된다면!
종단에는 문서를 읽지않고 코드를 읽는게 편하다고 느끼게 되는 개발자가 생길거에요!
그리고 문서를 읽지않고 코드를 읽겠죠!

JIY0UNG committed:
List.of()를 Arrays.asList()로 변경하라

List.of()는 자바 9부터 지원하는 메서드인데, 현재 프로젝트에서는 자바 1.8을 사용하도록 설정되어 있습니다. 따라서 자바 1.8에서 사용 가능한 Arrays.asList()를 사용하여, 자바 버전을 변경하지 않고도 List.of()와 같이 product를 리스트에 넣어줄 수 있게 수정했습니다.

JIY0UNG commented on app/src/test/java/com/codesoom/assignment/controllers/ProductControllerTest.java:
```diff
@@ -55,7 +55,7 @@ void setUp() {
                 .build();
 
-        given(productService.getProducts()).willReturn(List.of(product));
+        given(productService.getProducts()).willReturn(Arrays.asList(product));
```
https://github.com/CodeSoom/spring-week8-assignment-1/pull/62#issuecomment-1165574172 저도 같은 에러를 겪고, 승로님처럼 인텔리제이 jdk 버전을 변경해보기도 했는데 해결되지 않더라구요.
저 리뷰에서 영환님이 말씀하신 것 처럼 sourceCompatibility를 변경하지 않고 이 에러를 어떻게 해결할 수 있을 지 생각해봤습니다.

그 전에 sourceCompatibility가 무엇인지 살짝 알아봤습니다.
- sourceCompatibility는 자바 코드를 컴파일할 때 어떤 버전의 자바를 사용할 지 설정한다.
- sourceCompatibility에 설정된 자바 버전에 따라 소스 코드에서 사용할 수 있는 기능이 달라진다.

현재 sourceCompatibility가 1.8로 설정되어있으니 해당 버전에서 List.of()와 같은 동작을 하는 메서드로 변경해줘야겠다고 생각했고, Arrays.asList()로 바꿔줌으로써 에러를 해결할 수 있었습니다.

List.of()와 Arrays.asList()는 둘 다 리스트에 값을 넣어줄 때 사용하지만, 아래와 같은 차이가 있습니다.

Arrays.asList()
- 가변 리스트를 반환합니다.
- null을 허용합니다.
- 원본 리스트의 값이 변경되면 참조한 리스트의 값도 변경됩니다. 

List.of()
- 불변 리스트를 반환합니다.
- null을 허용하지 않습니다.
- 원본 리스트의 값이 변경되어도 참조한 리스트의 값은 변하지 않습니다.

---

참고
- [Gradle, "sourceCompatibility" vs "targetCompatibility"?](https://stackoverflow.com/questions/16654951/gradle-sourcecompatibility-vs-targetcompatibility)
- [What is the difference between List.of and Arrays.asList?](https://stackoverflow.com/questions/46579074/what-is-the-difference-between-list-of-and-arrays-aslist)

ssisksl77 replied:
좋습니다!! 지우지 않고 해결하는 방법 중에 코드를 변경하는 방법을 선택하셨군요!! ㅎㅎ

사실 저는 호환하는 버전을 1.8에서 11버전으로 변경하면 되지 않을까 라는 생각으로 문의드렸었습니다 ㅎㅎ;;

JIY0UNG committed:
MariaDB 의존성을 추가하라

JIY0UNG committed:
MariaDB profile을 설정하라

MariaDB 접속 정보에 대한 설정을 추가했습니다.
mariadb profile을 사용할 때에만 해당 설정을 따르도록 별도의 profile을 지정하였습니다.

JIY0UNG committed:
Dockerfile을 생성하라

이미지를 만들기 위해 필요한 명령어들을 Dockerfile에 작성하였습니다.
Dockerfile을 빌드 하여 생성된 이미지를 Docker 컨테이너로 실행하면, 서버가 구동됩니다.

JIY0UNG commented:
MariaDB를 사용하기 위한 설정을 했고, 애플리케이션 배포에 필요한 Dockerfile을 만들었습니다.
Dockerfile을 빌드하면 배포에 사용될 이미지가 생성되며, 이를 Docker 컨테이너로 실행하여 서버를 구동할 수 있습니다.

JIY0UNG commented:
어느새 8주 과정이 끝났네요.
그동안 해주신 리뷰들을 통해 많은 것들을 배울 수 있었어요.
정말 감사드립니다!!
